File: sdk-py/husk/_transport.py
"""JSON-RPC 2.0 client over HTTP for Husk SDK."""
from __future__ import annotations

import itertools
import json
from typing import Any, Mapping, Optional

import httpx


class JsonRpcTransportError(Exception):
    """Raised when the HTTP transport itself fails (non-200, bad JSON, etc.)."""


class HuskApiError(Exception):
    """Raised when the server returns a JSON-RPC error envelope."""

    def __init__(self, message: str, code: int, data: Any = None) -> None:
        super().__init__(message)
        self.code = code
        self.data = data


class JsonRpcClient:
    """Async JSON-RPC client. Reusable across many calls."""

    def __init__(
        self,
        base_url: str,
        *,
        http_client: Optional[httpx.AsyncClient] = None,
        timeout: float = 30.0,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._owns_client = http_client is None
        self._client = http_client or httpx.AsyncClient(timeout=timeout)
        self._ids = itertools.count(1)

    async def call(self, method: str, params: Mapping[str, Any]) -> Any:
        rpc_id = next(self._ids)
        url = f"{self._base_url}/v1/jsonrpc"
        try:
            res = await self._client.post(
                url,
                json={"jsonrpc": "2.0", "id": rpc_id, "method": method, "params": params},
            )
        except httpx.HTTPError as e:
            raise JsonRpcTransportError(f"HTTP transport failed: {e}") from e
        if res.status_code != 200:
            raise JsonRpcTransportError(f"HTTP {res.status_code} from {url}")
        try:
            body = res.json()
        except json.JSONDecodeError as e:
            raise JsonRpcTransportError("Response body was not valid JSON") from e
        if "error" in body:
            err = body["error"]
            raise HuskApiError(err["message"], err["code"], err.get("data"))
        return body["result"]

    async def aclose(self) -> None:
        if self._owns_client:
            await self._client.aclose()

    async def __aenter__(self) -> "JsonRpcClient":
        return self

    async def __aexit__(self, *_: Any) -> None:
        await self.aclose()

File: sdk-ts/src/transport.ts
import type { JsonRpcResponse } from "./types.js";

export class JsonRpcTransportError extends Error {
  constructor(message: string, public readonly cause?: unknown) {
    super(message);
    this.name = "JsonRpcTransportError";
  }
}

export class HuskApiError extends Error {
  constructor(
    message: string,
    public readonly code: number,
    public readonly data?: unknown
  ) {
    super(message);
    this.name = "HuskApiError";
  }
}

export interface JsonRpcClientOptions {
  /** e.g. "http://localhost:7777" — trailing slash is stripped. */
  baseUrl: string;
  /** Optional fetch override (for tests + custom transports). */
  fetch?: typeof globalThis.fetch;
}

/**
 * Minimal JSON-RPC 2.0 client over HTTP POST. v0 binds to /v1/jsonrpc on
 * the orchestrator. No retry, no batching, no timeout (caller handles).
 */
export class JsonRpcClient {
  private readonly _baseUrl: string;
  private readonly fetchImpl: typeof globalThis.fetch;
  private nextId = 0;

  constructor(opts: JsonRpcClientOptions) {
    this._baseUrl = opts.baseUrl.replace(/\/+$/, "");
    this.fetchImpl = opts.fetch ?? globalThis.fetch;
  }

  /** The orchestrator base URL (no trailing slash). Used by subscribe.ts for SSE. */
  get baseUrl(): string {
    return this._baseUrl;
  }

  /**
   * Invoke a JSON-RPC method. Returns the `result` payload on success or
   * throws `HuskApiError` (server returned `error`) or
   * `JsonRpcTransportError` (HTTP/parse issue).
   */
  async call<T = unknown>(method: string, params: Record<string, unknown>): Promise<T> {
    const id = ++this.nextId;
    const url = `${this._baseUrl}/v1/jsonrpc`;
    let res: Response;
    try {
      res = await this.fetchImpl(url, {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify({ jsonrpc: "2.0", id, method, params }),
      });
    } catch (e) {
      throw new JsonRpcTransportError(`Fetch failed: ${(e as Error).message}`, e);
    }
    if (!res.ok) {
      throw new JsonRpcTransportError(`HTTP ${res.status} from ${url}`);
    }
    let body: JsonRpcResponse<T>;
    try {
      body = (await res.json()) as JsonRpcResponse<T>;
    } catch (e) {
      throw new JsonRpcTransportError("Response body was not valid JSON", e);
    }
    if ("error" in body) {
      throw new HuskApiError(body.error.message, body.error.code, body.error.data);
    }
    return body.result;
  }
}
